packages/core/src/browser/label-provider.ts

```diff
@@ -48,6 +49,11 @@ export interface LabelProviderContribution {
      */
     getLongName?(element: object): string;
 
+    /**
+     * Emit when something has changed that may result in this label provider returning a different
+     * value for one or more properties (name, icon etc) for the emitted URI.
+     */
+    readonly onDidChange?: Event<URI>;
```

Comment:
I still don't like that it is `URI`, it should support other element to be aligned with other method, how about:
```
export interface DidChangeLabelEvent {
    affects(element: object): boolean;
}
```

Under the hood it should unwrap/down-case given element similar to other methods.

Reply:
The label provider contributions allow conversions before delegating to lower priority contributions.  This makes sense as it allows, for example, providers be given extra information to the URI such as the folder/file flag in a FileStat.  However these all reduce down to a URI and that is all a UI component should need to identify what is being updated.  It would seem to me to be more complex if UI components could get anything.  How could that work?  A UI component would then need to know how to get a URI for any object,

Reply:
It is not clear to me how these unwrap/down-case are managed.  For example, WorkspaceUriLabelProviderContribution down cases from (URI | FileStat) to URI.  However it passes the URI on the super, being DefaultUriLabelProviderContribution.  However the priority of WorkspaceUriLabelProviderContribution is 10 and DefaultUriLabelProviderContribution is 1 so all contributions with priorities 2 through 9 are ignored.  It is not clear to me if this is by design or is a bug.  It would be easy enough to pass on to the next highest priority, instead of extending DefaultUriLabelProviderContribution but it is then not clear to me how label contributors are to know what types they may get and what types they may pass on to the lower priority providers.

Reply:
> A UI component would then need to know how to get a URI for any object

An example of the problem here can be seen in the navigator.  If we want to be able to lookup the node in the `nodes` map then we need to be able to get the key.  Perhaps it is ok to scan the entire map testing each with `affects`, otherwise we are going to need the URI to calculate the map key.  However I can see the cleanliness of calling `affects` for every possible node in the UI component and I assume you are ok with the possible performance loss.  On the plus side, having the `affects` callback would be more efficient if a large number of labels needed refreshing at once.  So I think we are ok, the more I think about it your suggestion is going to work well.

Reply:
You are right performance issue will be real for the navigator.

An idea of `LabelProvider` was that it can give a system-wide UI representation for some object, it does not need to be URI at all. A contributor implements conversions and then clients can drop an object into `LabelProvider` and get proper values.

Maybe we can have a compromised solution for an optimization, i.e. to add `uri` optional property to `DidChangeLabelEvent`? The navigator can check whether `uri` is not defined and use it.

Reply:
I thought about it, and i think we can rather introduce a subtype for `DidChangeLabelEvent` for uri-based elements, i.e.
```
export interface DidChangeUriLabelEvent extends DidChangeLabelEvent {
    uri: URI
}
export namespace DidChangeUriLabelEvent {
    export function is(event: DidChangeLabelEvent): event is DidChangeUriLabelEvent {
        return 'uri' in event && event['uri'] instanceof URI;
    }
}
```

So a navigator just checks that an event is of `DidChangeUriLabelEvent` type and use `uri`.

Reply:
When I looked at the navigator, I saw around 250 nodes on my portrait mode monitor after a short amount of activity opening and closing nodes.   I changed my mind and decided that testing all of 250 nodes is not a performance issue and it keeps things simpler.

Reply:
I usually traverse node_modules for a bit in Theia repo and then test such things. It is typical that a user navigates to a library and end up loading a lot of nodes from node_modules.